File: pc-square-io-separator-java/src/main/java/com/mj/pc/square/io/separator/java/exception/ExecutionException.java
package com.mj.pc.square.io.separator.java.exception;

import lombok.NoArgsConstructor;

import java.util.Optional;
import java.util.ResourceBundle;

/**
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 28-10-2022
 */

@NoArgsConstructor
public abstract class ExecutionException extends java.util.concurrent.ExecutionException {

    private static final ResourceBundle resources = ResourceBundle.getBundle("exceptions");

    public ExecutionException(String message) {
        super(message);
    }

    protected abstract String getExceptionKey();

    @Override
    public final String getMessage() {
        return Optional
                .ofNullable(getExceptionKey())
                .map(resources::getString)
                .orElse(null);
    }
}

File: pc-square-io-separator-kotlin/src/main/kotlin/com/mj/pc/square/io/separator/kotlin/exception/ExecutionException.kt
package com.mj.pc.square.io.separator.kotlin.exception

import lombok.NoArgsConstructor
import java.util.Optional
import java.util.ResourceBundle
import java.util.concurrent.ExecutionException

/**
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 28-10-2022
 */

@NoArgsConstructor
abstract class ExecutionException : ExecutionException {

    companion object {

        private val resources = ResourceBundle.getBundle("exceptions")
    }

    constructor() : super()

    constructor(message: String?) : super(message)

    protected abstract val exceptionKey: String?

    override val message: String?
        get() = Optional.ofNullable(exceptionKey).map(resources::getString).orElse(null)
}